export/mlab_export.py

```diff
@@ -0,0 +1,591 @@
+    msg = ('Start and end times difference is less than length: '+
+           '%s - %s = %s < %s' % (options.ts_end, options.ts_start,
+           (options.ts_end - options.ts_start), options.length))
+    raise TimeOptionError(msg)
+
+  logging.debug('Exporting: %s to %s', time.ctime(options.ts_start),
+                time.ctime(options.ts_end))
+
+
+def default_output_name(ts_start, ts_end, output_dir):
+  """Creates a default output filename based on time range and output dir.
+
+  Filenames are formatted with time stamps as:
+      <output_dir>/resource-utilization/YYYY/MM/DD/<HOSTNAME>/
+          metrics-YYYYMMDDTHHMMSS-to-YYYYMMDDTHHMMSS.json
+
+  Args:
+    ts_start: int, starting timestamp of export in seconds since the epoch.
+    ts_end: int, ending timestamp of export in seconds since the epoch.
+    output_dir: str, base path of directory for output.
+  Returns:
+    str, absolute path of generated output file name.
+  """
+  filename = 'metrics-%s-to-%s.json' % (
+      time.strftime('%Y%m%dT%H%M%S', time.localtime(ts_start)),
+      time.strftime('%Y%m%dT%H%M%S', time.localtime(ts_end)))
+  date_path = time.strftime('%Y/%m/%d', time.localtime(ts_start))
+  full_path = os.path.join(
+      output_dir, 'resource-utilization', date_path, HOSTNAME)
+  if not os.path.exists(full_path):
+    os.makedirs(full_path)
+  return os.path.join(full_path, filename)
+
+
+def get_canonical_names(filename, value_name, options):
+  """Converts raw filename and value names from RRD into canonical export names.
+
+  Args:
+    filename: str, the absolute path of an rrd file.
+    value_name: str, the name of the value being exported from the RRD.
+    options: flags.FlagValues, the runtime options. This method uses
+        options.rrddir_prefix and all option.show_* flags.
+  Returns:
+    (str, str, str), with HOSTNAME, experiment name, metric name.
+  """
+  # Strip rrddir_prefix and split directory components.
+  file_fields = filename.replace(options.rrddir_prefix, '').split('/')
+
+  # The zeroth field is always the context hostname.
+  if HOSTNAME == file_fields[0]:
+    # The root context represents whole-system metrics.
+    experiment = 'system'
+  else:
+    # A slice hostname. Everything remaining after stripping hostname.
+    experiment = file_fields[0].replace('.' + HOSTNAME, '')
+
+  raw_fields = file_fields[1:-1]
+  raw_fields.append(file_fields[-1].replace('.rrd', ''))
+  if value_name != 'value':
+    raw_fields.append(value_name)
+
+  metric_raw = '.'.join(raw_fields)
+
+  # NOTE: convert the raw metric name into the canonical form, or None.
+  metric = METRIC_MAP.get(metric_raw, None)
+
+  # Optionally print extra information.
+  if options.show_nagios:
+    cmd = ('collectd-nagios -s $COLLECTD_UNIXSOCK -H {host} '+
+           '-n {metric} -d {value} [-w <l:h> -c <l:h>]')
+    cmd = cmd.format(host=file_fields[0],
+                     metric='/'.join(file_fields[1:]).replace('.rrd', ''),
```

Comment:
Reviewing the logic in this method again myself, I had a hard time. Trying to answer your question about raw_fields was overly difficult. So, I've tried to simplify it here: 2ac42051282629edf24f67cdf9f5f6dae06bc968

How does it looks?